plugins/src/gcc/imp.rs

```diff
@@ -0,0 +1,1360 @@
+                self.increasing_counter += 1;
+
+                if self.increasing_duration > *OVERUSE_TIME_TH
+                    && self.increasing_counter > 1
+                    && estimate > self.last_overuse_estimate
+                {
+                    self.usage = NetworkUsage::Over;
+                }
+            }
+            NetworkUsage::Under | NetworkUsage::Normal => {
+                self.increasing_duration = Duration::zero();
+                self.increasing_counter = 0;
+
+                self.usage = th_usage;
+            }
+        }
+        self.last_overuse_estimate = estimate;
+    }
+}
+
+#[derive(Default, Debug)]
+struct ExponentialMovingAverage {
+    average: Option<f64>,
+    variance: f64,
+    standard_dev: f64,
+}
+
+impl ExponentialMovingAverage {
+    fn update<T: Into<f64>>(&mut self, value: T) {
+        if let Some(avg) = self.average {
+            let avg_diff = value.into() - avg;
+
+            self.variance = (1. - MOVING_AVERAGE_SMOOTHING_FACTOR)
+                * (self.variance + MOVING_AVERAGE_SMOOTHING_FACTOR * avg_diff * avg_diff);
+            self.standard_dev = self.variance.sqrt();
+
+            self.average = Some(avg + (MOVING_AVERAGE_SMOOTHING_FACTOR * avg_diff));
+        } else {
+            self.average = Some(value.into());
+        }
+    }
+
+    fn estimate_is_close(&self, value: Bitrate) -> bool {
+        self.average.map_or(false, |avg| {
+            ((avg - STANDARD_DEVIATION_CLOSE_NUM * self.standard_dev)
+                ..(avg + STANDARD_DEVIATION_CLOSE_NUM * self.standard_dev))
+                .contains(&(value as f64))
+        })
+    }
+}
+
+struct State {
+    /// Note: The target bitrate applied is the min of
+    /// target_bitrate_on_delay and target_bitrate_on_loss
+    estimated_bitrate: Bitrate,
+
+    /// Bitrate target based on delay factor for all video streams.
+    /// Hasn't been tested with multiple video streams, but
+    /// current design is simply to divide bitrate equally.
+    target_bitrate_on_delay: Bitrate,
+
+    /// Used in additive mode to track last control time, influences
+    /// calculation of added value according to gcc section 5.5
+    last_increase_on_delay: Option<time::Instant>,
+    last_decrease_on_delay: time::Instant,
+
+    /// Bitrate target based on loss for all video streams.
+    target_bitrate_on_loss: Bitrate,
+
+    last_increase_on_loss: time::Instant,
+    last_decrease_on_loss: time::Instant,
+
+    /// Exponential moving average, updated when bitrate is
+    /// decreased
+    ema: ExponentialMovingAverage,
+
+    last_control_op: BandwidthEstimationOp,
+
+    min_bitrate: Bitrate,
+    max_bitrate: Bitrate,
+
+    detector: Detector,
+
+    clock_entry: Option<gst::SingleShotClockId>,
+
+    // Implemented like a leaky bucket
+    buffers: VecDeque<gst::Buffer>,
+    // Number of bits over the budget sent during last burst
+    last_overshot_bits: u64,
```

Comment:
fairly sure making this an `i64` would simplify the code ;)

Reply:
It should **never** be negative.

Reply:
what I'm saying is you could reverse the meaning of the field, add it to the budget instead of remove it, and the code will be cleaner looking ..

Reply:
Sensibly, but I can't find a good name for it, in practice that is the change I could make:

``` diff
diff --git a/plugins/src/gcc/imp.rs b/plugins/src/gcc/imp.rs
index 3b3bbd3..69c775f 100644
--- a/plugins/src/gcc/imp.rs
+++ b/plugins/src/gcc/imp.rs
@@ -722,8 +722,8 @@ struct State {
 
     // Implemented like a leaky bucket
     buffers: VecDeque<gst::Buffer>,
-    // Number of bits over the budget sent during last burst
-    last_overshot_bits: u64,
+    // Number of bits over the budget sent during last burst, always <= 0
+    last_overshot_bits: i64,
 
     flow_return: Result<gst::FlowSuccess, gst::FlowError>,
     last_push: time::Instant,
@@ -763,7 +763,7 @@ impl State {
                 self.estimated_bitrate as i64,
                 gst::ClockTime::SECOND.nseconds() as i64,
             )
-            .unwrap() - self.last_overshot_bits as i64;
+            .unwrap() + self.last_overshot_bits;
         let total_budget = budget;
         let mut remaining = self.buffers.iter().map(|b| b.size() as f64).sum::<f64>() * 8.;
         let total_size = remaining;
@@ -799,7 +799,7 @@ impl State {
         );
 
         self.last_push = now;
-        self.last_overshot_bits = if budget < 0 && !leaked { -budget as u64 } else { 0 };
+        self.last_overshot_bits = if !leaked { budget } else { 0 };
 
         list
     }
 ```

I don't think it is worth it.